File: Programmers/[42891] 무지의 먹방 라이브/woong-jae/42891.js
function solution(food_times, k) {
    const sorted_food_times = food_times.slice(0).sort((a, b) => a - b);
    
    let ate = 0, i = 0;
    for(i; i < food_times.length; i++) {
        const leftFood = food_times.length - i;
        const neededTime = (sorted_food_times[i] - ate) * leftFood;
        if(neededTime > k) break;
        ate += (sorted_food_times[i] - ate);
        k -= neededTime;
    }
    if(i === food_times.length) return -1;

    const reduced_food_times = [];
    food_times.forEach((time, index) => {
        if(time <= ate) return;
        reduced_food_times.push([time - ate, index + 1]);
    });
    
    return reduced_food_times[k % reduced_food_times.length][1];
}

File: Programmers/[42891] 무지의 먹방 라이브/ONE/42891.java
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;

class Food{
    int num;
    int time;

    public Food(int num, int time) {
        this.num = num;
        this.time = time;
    }
}

class Solution {
    public int solution(int[] food_times, long k) {
        List<Food> list = new ArrayList<>();
        int n = food_times.length;

        for (int i = 1; i <= n; i++)
            list.add(new Food(i, food_times[i - 1]));

        list.sort(Comparator.comparingInt(o -> o.time));

        int index = 0;
        int previous = 0;
        for (Food current : list) {
            long diff = current.time - previous;

            if (diff != 0) {
                long spend = diff * n;

                if (spend <= k) {
                    k -= spend;
                    previous = current.time;
                }
                else {
                    k %= n;
                    list.subList(index, list.size()).sort(Comparator.comparingInt(o -> o.num));
                    return list.get(index + (int) k).num;
                }
            }
            ++index;
            --n;
        }
        return -1;
    }
}
